docs/exceptions.md

```diff
@@ -155,6 +181,11 @@ Do:
 * use `MonadError` or `MonadThrow` (methods of these classes). Note
   that if you define `f :: MonadError m => m ()`, it won't be pure
 
+*Comment:* I disagree with the last "Do not". Why is this ruled out? Sometimes
+giving `Either Text` is legit. Giving `Maybe` is ok, so it must also ok to
+give `Either ()`, so why not `Either Text`, which contains even more
```

Comment:
The point is that if you're going to give more information, do it via a dedicated AST, not via `Text`, because it will simplify the life of anyone who might ever need to catch this error.

Reply:
to illustrate, 
```haskell
case someValue of
  Left "this error happened" -> ...
```
is not a good way to handle this

Reply:
It lies between two points:

```Haskell
case soemValue of
  Left MyExpressiveErrorConstructor -> ...
```

and

```Haskell
case someValue of
    Nothing -> ...
```

Why are these 2 sanctioned, but using `Text` is not?

Reply:
For the record: I agree that using a custom type is better than `Text`, I just think we're looking at dogma here.

Reply:
`Nothing` implies "there is exactly one way this could have failed," and that's IMO a suitable type for something like `Map.lookup`

`Either` implies "there are multiple ways this could fail, *and* you want to be able to handle all the possible failures." That suggests to me that you should document these possible failure modes you're expecting, and allow programmers to handle them.

Reply:
> Nothing implies "there is exactly one way this could have failed," and that's IMO a suitable type for something like Map.lookup

Careful: the lookup didn't fail. It succeeded. Nothing went wrong. There was no key at that value.

> Either implies "there are multiple ways this could fail, and you want to be able to handle all the possible failures." That suggests to me that you should document these possible failure modes you're expecting, and allow programmers to handle them.

Does it? What if there's precisely one way it could fail, and it comes with a `Text`?

Reply:
> Careful: the lookup didn't fail. It succeeded. Nothing went wrong. There was no key at that value.

If the goal of the lookup is to retrieve the value, not having the value in the map is a failure. Of course, this failure is represented as a normal value, `Nothing`, and it's not something that would crash our program, it's a different sort of failure.

Reply:
>What if there's precisely one way it could fail, and it comes with a Text?

I guess that's a bad interface design then. Document your `foo :: a -> Maybe b` so that it'd be clear: if you get `Nothing` then "exactly X went wrong".

Reply:
> If the goal of the lookup is to retrieve the value, not having the value in the map is a failure. Of course, this failure is represented as a normal value, Nothing, and it's not something that would crash our program, it's a different sort of failure.

There's no "goal" of lookup. What's the "goal" of `(+) 1`? What's the "goal" of an integral? What's the "goal" of the expected value of a random variable? These are mathematical functions. They don't have objectives, they just *are*.

Reply:
> I guess that's a bad interface design then. Document your foo :: a -> Maybe b so that it'd be clear: if you get Nothing then "exactly X went wrong".

And if I have a function which returns `Either Text Foo`, where the `Left txt` does *not* subjectively reflect an "unwanted" "code path"? Is that OK?

Reply:
> There's no "goal" of lookup. What's the "goal" of (+) 1? What's the "goal" of an integral? What's the "goal" of the expected value of a random variable? These are mathematical functions. They don't have objectives, they just are.

Programmers have objectives when they write code. The code itself doesn't have any goals of course, but it reflects the goals the programmer had. Maybe this is a bad choice of words on my part, I should've said "purpose", not "goal". 

> And if I have a function which returns Either Text Foo, where the Left txt does not subjectively reflect an "unwanted" "code path"? Is that OK?

This is ok because

* `Text` has no `Exception` instance (this holds)
* `Text` is not an error message of any sort (otherwise I'd say it does reflect an "unwanted" "code path")
* the use of `Either` is symmetric here, so if we do `Either Foo Text`, nothing will change

But assume we use the `Monad` instance for `Either` that gives a special treatment to the `Left` constructor, using it to short-circuit the computation with a "failure". In this case using `Text` wouldn't be OK.